### PR TITLE
Enable "useProvidedKeypoints" flag on cuda::ORB's detectAndCompute

### LINKED ISSUE #6473: [Question] Why cuda::ORB detectAndCompute's  useProvidedKeypoints parameter is forced to be false?
Hi,
I would like to use cuda::orb compute or detectAndCompute method to describe keypoints that I detect earlier. But when I try to use these methods I got "Assertion failed (useProvidedKeypoints == false) in detectAndCompute" error. When I search on the source code, I see 
```
void ORB_Impl::detectAndCompute(InputArray _image, InputArray _mask, std::vector<KeyPoint>& keypoints, OutputArray _descriptors, bool useProvidedKeypoints){

CV_Assert( useProvidedKeypoints == false ); 
...
```
Why does the function takes a useProvidedKeypoints parameter that is forced to be equal to false?

Is there another way to achieve my goal?
### my system
- OpenCV version:  3.1
- Host OS: Linux (Ubuntu 14.04) 

- cuda::orb, cuda, orb

### Code example 
```
    cv::cuda::GpuMat gpu_img;
    cv::cuda::GpuMat dst;

    cv::Mat img = cv::imread("Lena.pgm", CV_LOAD_IMAGE_GRAYSCALE);
    gpu_img.upload(img);

    cv::Ptr<cv::cuda::ORB> orb = cv::cuda::ORB::create();
    std::vector<cv::KeyPoint> keypoints;
    orb->detect(gpu_img, keypoints); // reason not to use detectAndCompute method only, I detect keypoints earlier and do same calculations to produce new keypoints based on old ones, then I need to describe new ones.   
 orb->compute(gpu_img, keypoints,dst); 
// same with orb->detectAndCompute(gpu_img, noArray(), keypoints,dst, true);
```

### PR DESCRIPTION
**Pull request by OzgrCn**
resolves #6473

Enable "useProvidedKeypoints" flag on cuda::ORB's detectAndCompute function in order to able to describe provided keypoints.
 * Seperate keypoints according to their octave
 * Rearrange their merged locations based on scale factor.
 * Upload keypoints into GPU
 * Describe them

* additional author : Ender Tunç <endertunc@yahoo.com>

### CODE REVIEW

File: modules/cudafeatures2d/src/orb.cpp [hunk 1]
```diff
@@ -570,30 +570,92 @@ namespace
...
+                level = keypoints[j].octave;
+                oKeypoints[level].push_back(keypoints[j]);
+            }
+            if (!keypoints.empty())
+            {
+                keyPointsPyr_.resize(nLevels_);
+                keyPointsCount_.resize(nLevels_);
+                int t;
+                for(t = 0; t < nLevels_; t++) {
+                    std::vector<KeyPoint> ks = oKeypoints[t];
```
Comment (jet47) [inline modules/cudafeatures2d/src/orb.cpp]: Please use const reference `const std::vector<KeyPoint>& ks` to avoid array copy on each iteration.

**Comment by OzgrCn** [comment]
@jet47 Done

File: modules/cudafeatures2d/src/orb.cpp [hunk 2]
```diff
@@ -604,7 +604,7 @@ namespace
                 keyPointsCount_.resize(nLevels_);
                 int t;
                 for(t = 0; t < nLevels_; t++) {
-                    std::vector<KeyPoint> ks = oKeypoints[t];
+                    const std::vector<KeyPoint> ks = oKeypoints[t];
```
Comment (jet47) [inline modules/cudafeatures2d/src/orb.cpp]: It should be const *reference* (you missed `&`).

```
const std::vector<KeyPoint>& ks = oKeypoints[t];
```

Comment (OzgrCn) [inline modules/cudafeatures2d/src/orb.cpp]: @jet47 Right, I missed it, thanks. I have updated it.

**Comment by alalek** [comment]
Please squash all commits into one also.

**Comment by jet47** [comment]
:+1:

**Comment by OzgrCn** [comment]
@alalek @jet47 Ok, I have squashed.

**Pull request merged**